### DISCUSSION TITLE
Functions in html called before init() finishes

### DISCUSSION THREAD

**Discussion by AndiLeni** [alpinejs/alpine#4226] [1. Help] [answered]
Hello,

I noticed, that functions which are called from html are executed before the init() function finishes.
This is bad in my case, since I load data from the server which is later shown on the page.

https://jsitor.com/akZ_VIpUQZm

Html:
```html
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>

<body x-data="app()">

	<span class="fw-bold" x-text="getConvNameById(12)"></span>

</body> 
```

JS:
```javascript
function app() {
    return {
        posts: [],

        async init() {
            console.log("initializing...")

            posts = await fetch('https://jsonplaceholder.typicode.com/posts');
            console.log(posts);
            this.posts = posts;

            console.log("initialized")
        },

        getConvNameById(id) {
            console.log(id)
        }
    }
}
```

This logs the following to the console:
```
initializing...
12
Response { … }
initialized
```

But I would have expected this behavior:
```
initializing...
Response { … }
initialized
12
```

It seems to be caused by making init() async, but it is counter inituitive since I would have expected alpine to wait nonethless until init() finishes.

Kind regards
Andreas

**Comment by SimoTod** [comment, marked as answer]
It's how javascript works. Asynchronous means that the other functions can run before the function is returned.
It's not an Alpine thing. 

You can have something like this
```javascript
document.addEventListener('alpine:init', () => Alpine.data('app', () => {
    return {
        loaded: false
        posts: [],

        async init() {
            console.log("initializing...")

            posts = await fetch('https://jsonplaceholder.typicode.com/posts');
            console.log(posts);
            this.posts = posts;
            this.loaded = true;

            console.log("initialized")
        },

        getConvNameById(id) {
            console.log(id)
        }
    }
}))
```

```html
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
<body x-data="app"> 
  <template x-if="loaded">
    <span class="fw-bold" x-text="getConvNameById(12)"></span>
  </template>
</body>
```

**Reply by AndiLeni** [reply]
Thanks, I will consider doing it the way you suggested. 
I expected alpine to await the init function in all cases, but seems like I had false assumptions.

**Reply by SimoTod** [reply]
Yeah, I don't think javascript allows you to make an asynchronous operation synchronous so Alpine can't do much on that point

**Reply by ekwoka** [reply]
Yup, and you'd never really want to anyway.

**Comment by ekwoka** [comment]
Yup, as above, Alpine doesn't wait for async directives and stuff. And 99% of the time ,having it await the init would not be what people want.

Also, fyi, `posts = ` is assigning a global variable (and would be an error in strict mode). Probably shouldn't do that.

**Comment by boustanihani** [comment]
AKAIK `x-init` awaits async functions by default. 

Wouldn't it be good for consistency, if the **init()** function inside **Alpine.data** is also awaited when asynchronous?

**Reply by ekwoka** [reply]
I think the issue there is that the tree is handled synchronously, awaiting the init like that could cause a kind of waterfall loading thing for nested components. Of course, it would be available in userland to just call an async method in the init to avoid it, but I'm also not sure what is the expected issue it would solve. Is there a reason it should actually wait and not initialize everything?

**Reply by boustanihani** [reply]
IMO, it has some positive sides: 
- It would eliminate the need to define default values (like empty array `[]` boilerplate), and this will result in cleaner code.
- No more `Uncaught ReferenceError` (not defined errors) when a default value is missing 
- For consistency, I would expect Alpine to detect when `init()` is declared as async and automatically await, which is the same behavior for x-init 

When a component has to be awaited, it could be added to a waiting stack, and other components can be a rendered in that time, so no performance issues occur.

**Reply by ekwoka** [reply]
it would still cause delays since components can be inside other components.

> It would eliminate the need to define default values (like empty array [] boilerplate), and this will result in cleaner code.

I disagree, now the code is less clear, not more clear.

You're encouraged to explicitly declare the properties you'll use on those components, since the proxy stack could find that key on a parent and now your data is f'd up.

Include the keys where you want them, so the framework and other devs know where it lives.

**Comment by boustanihani** [comment]
> You're encouraged to explicitly declare the properties you'll use on those components, since the proxy stack could find that key on a parent and now your data is f'd up.
> 
> Include the keys where you want them, so the framework and other devs know where it lives.

I agree, this is good practice, but what I dislike is the inconsistent behavior when using async functions. 

Take the following code:

```html
<!DOCTYPE html>

<html lang="en">

<head>
    <meta charset="utf-8" />
    <title>x-init vs init()</title>
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3/dist/cdn.min.js"></script>
</head>

<body>

    <h3>Userlist-1 using x-init:</h3>
    <ul x-data x-init="userlist1=fetchUsers()">
        <template x-for="user in userlist1">
            <li x-text="user"></li>
        </template>
    </ul>

    <h3>Userlist-2 using init():</h3>
    <ul x-data="userlist2">
        <template x-for="user in userlist2">
            <li x-text="user"></li>
        </template>
    </ul>

    <script>

        const sleep = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

        async function fetchUsers() {
            await sleep(1000);
            return ["user1", "user2", "user3"];
        }

        document.addEventListener('alpine:init', () => {

            Alpine.data('userlist2', () => ({

                // userlist2: [], // Solves: "Uncaught ReferenceError: userlist2 ..."

                async init() { // Is not getting awaited
                    this.userlist2 = await fetchUsers();
                }
            }));
        });

    </script>

</body>

</html>
```

`Userlist-1` is populated fine without any errors on the console. 

`Userlist-2 `causes `Uncaught ReferenceError: userlist2 is not defined` on first render, because `async init()` is not awaited. 

<img width="1150" height="326" alt="Bildschirmfoto vom 2026-02-23 13-15-02" src="https://github.com/user-attachments/assets/25c635f2-bd69-43e8-9638-67bbc9aa223e" />

Uncommenting `userlist2: []` solves the problem. 

I am only suggesting that `init()` functions inside `Alpine.data` should also get awaited when `async`.

Here is the whole html file: [alpine-init-behaviour.html](https://github.com/user-attachments/files/25484270/alpine-init-behaviour.html)
